## OR 506: HomeWork-4
**Created in iPython Notebook by Guduguntla Vamshi <gudugu@ncsu.edu>**

##  1.Conjugate Gradient Method:

\\begin{equation}
min \ f(x),\ f(x) = 0.5x^T A x - b^T x
\\end{equation}

\begin{eqnarray}
\ A = \begin{pmatrix}
4 & 0 & 0 & 1 & 0 & 0\\
4 & 4 & 0 & 0 & 1 & 0 \\
0 & 0 & 5 & 0 & 0 & 1 \\
1 & 0 & 0 & 5 & 0 & 0 \\
0 & 1 & 0 & 0 & 6 & 0 \\
0 & 0 & 1 & 0 & 0 & 6 \\
\end{pmatrix}
\end{eqnarray}

\\begin{equation}
\ Using \ b^T = [4,-8,16,1,-2,9]
\\end{equation}

In [2]:
import math
import numpy as np

**Initializing the vectors**

In [3]:
x0 = np.matrix([[0],[0],[0],[0],[0],[0]])
A  = np.matrix([[4,0,0,1,0,0],[0,4,0,0,1,0],[0,0,5,0,0,1],
                [1,0,0,5,0,0],[0,1,0,0,6,0],[0,0,1,0,0,6]]) 
b =  np.matrix([[4],[-8],[16],[1],[-2],[9]])

**The Function to compute the minimizer vector by Conjugate gradient method**

In [4]:
def conjugate_gradient(A, b, x0, tol = 1.0e-8, max_iter = 100):
    """
    A function to solve [A]{x} = {b} linear equation system with the 
    conjugate gradient method.
    
    :param A : array 
        A real symmetric positive definite matrix(assumed)
        
    :param b : vector
        The vector of the system which is given in RHS.
        
    :param x0 : vector
        The starting guess for the solution.
        
    :param max_iter : integer
        Maximum number of iterations. Iteration will stop after 
        max_iter steps even if the specified tolerance 
        has not been achieved.
        
    :param tol : float
        Tolerance to achieve. The algorithm will terminate when either 
        the relative or the absolute residual is below tol.
        
    :var    r0 : vector
                 Initialization stores the value (b - a * A )
    
    :var    d  : vector
    
    :var    a  : float
                 Iteratively computes the 
                 scalar of (r1T.r1)/(r0T.r0)

    :var    ri : vector
                 Iteratively stores the value (r - a * A * d), 
                 used to check for the convergence
    
    :var    x  : vector 
                 Stores the solution for the next iteration iteratively
                 
    :var    b  : float
                 Iteratively computes the scalar of (riT.ri)/(diT.A.di)
    """
    x = x0
    r0 = b - np.dot(A, x)
    d = r0

#   Iterations:   
    for i in xrange(max_iter):
        a = float(np.dot(r0.T, r0)/np.dot(np.dot(d.T, A), d))
        x = x + d*a
        ri = r0 - np.dot(A*a, d)
        
        print "iteration: ",i, "r(i): ",round(np.linalg.norm(ri),5)

        if np.linalg.norm(ri) < tol:
            print "\nConverged Successfully in iterations :",i
            print "The result of vector x:"
            return np.around(x,decimals=10)
            break
        b = float(np.dot(ri.T, ri)/np.dot(r0.T, r0))
        d = ri + b * d
        r0 = ri
    return x

In [5]:
conjugate_gradient(A, b, x0, tol = 1.0e-10, 
                   max_iter = 100)

iteration:  0 r(i):  20.54264
iteration:  1 r(i):  10.48844
iteration:  2 r(i):  5.69484
iteration:  3 r(i):  3.39053
iteration:  4 r(i):  2.18629
iteration:  5 r(i):  1.41935
iteration:  6 r(i):  0.86272
iteration:  7 r(i):  0.48876
iteration:  8 r(i):  0.27119
iteration:  9 r(i):  0.15636
iteration:  10 r(i):  0.09683
iteration:  11 r(i):  0.06273
iteration:  12 r(i):  0.03945
iteration:  13 r(i):  0.02287
iteration:  14 r(i):  0.01234
iteration:  15 r(i):  0.00647
iteration:  16 r(i):  0.00345
iteration:  17 r(i):  0.00197
iteration:  18 r(i):  0.00123
iteration:  19 r(i):  0.00083
iteration:  20 r(i):  0.00055
iteration:  21 r(i):  0.00034
iteration:  22 r(i):  0.0002
iteration:  23 r(i):  0.00011
iteration:  24 r(i):  6e-05
iteration:  25 r(i):  4e-05
iteration:  26 r(i):  2e-05
iteration:  27 r(i):  1e-05
iteration:  28 r(i):  1e-05
iteration:  29 r(i):  0.0
iteration:  30 r(i):  0.0
iteration:  31 r(i):  0.0
iteration:  32 r(i):  0.0
iteration:  33 r(i):  0.0
iteration:  34 r(i)

array([[ 1.],
       [-2.],
       [ 3.],
       [ 0.],
       [-0.],
       [ 1.]])

**Comments on Results:**

- Convergence criteria is: 100 maximum iterations or **||ri|| ≤ 10^(-8)**
- Using the initial guess as xT = [0,0,0,0,0,0] approxiamated value of x:
  - is **[1.0,-2.0,3.0,0,0,1.0]**
- The algorithm converged in **49 iterations** using the above guess vector
- The computed values are cross-validated:
   - by plugging in these values given by the criteria defined above 
   - and by checking for convergence ~0.

**Comment on Method Used:**
- The conjugate gradient method works by generating a set of vectors d :
   - that are conjugate with respect to the matrix A. 
   - that is, **dTi A dj = 0, i != j**
- The formula for αi( used as **a** here) corresponds to an:
  - **exact line search along the direction di(used as d in     the code)**

##  2.Non-Linear Conjugate Gradient Method:

***Minimize***


$min \ f(x),\ f(x) = \sum\limits_{i=1}^{n-1} (x_i - 2x^2_{i+1})^2$


$\ x_0 = [1,1,1,1]^T$

In [7]:
import math
import numpy as np
from sympy import *

**Initializing the vectors**

In [9]:
x0 = np.array([1,1,1,1])

In [10]:
#function Defition

def func(a,b,c,d):
    x,y,z,w=symbols('x y z w')
    f = (x - 2*y**2)**2 + (y - 2*z**2)**2 + (z - 2*w**2)**2
    return f.subs({x:a,y:b,z:c,w:d})

def derv_x(a,b,c,d):
    x,y,z,w=symbols('x y z w')
    f = (x - 2*y**2)**2 + (y - 2*z**2)**2 + (z - 2*w**2)**2
    return diff(f,x).subs({x:a,y:b,z:c,w:d})
    
def derv_y(a,b,c,d):
    x,y,z,w=symbols('x y z w')
    f = (x - 2*y**2)**2 + (y - 2*z**2)**2 + (z - 2*w**2)**2
    return diff(f,y).subs({x:a,y:b,z:c,w:d})    
    
def derv_z(a,b,c,d):
    x,y,z,w=symbols('x y z w')
    f = (x - 2*y**2)**2 + (y - 2*z**2)**2 + (z - 2*w**2)**2
    return diff(f,z).subs({x:a,y:b,z:c,w:d})
    
def derv_w(a,b,c,d):
    x,y,z,w=symbols('x y z w')
    f = (x - 2*y**2)**2 + (y - 2*z**2)**2 + (z - 2*w**2)**2
    return diff(f,w).subs({x:a,y:b,z:c,w:d})

def grad_vector(x0):
    return np.matrix([derv_x(x0[0],x0[1],x0[2],x0[3]),
                      derv_y(x0[0],x0[1],x0[2],x0[3]),
                      derv_z(x0[0],x0[1],x0[2],x0[3]),
                      derv_w(x0[0],x0[1],x0[2],x0[3])])
def grad_vector1(x0):
    return np.array([derv_x(x0[0],x0[1],x0[2],x0[3]),
                     derv_y(x0[0],x0[1],x0[2],x0[3]),
                     derv_z(x0[0],x0[1],x0[2],x0[3]),
                     derv_w(x0[0],x0[1],x0[2],x0[3])])

In [11]:
grad_vector(x0) # gradient vector at x0 - guess vector

matrix([[-2, 6, 6, 8]], dtype=object)

In [13]:
def non_linear_conjugate_gradient(x0, tol = 1.0e-8, max_iter = 4):
    """
    A function to solve [A]{x} = {b} linear equation system with the 
    conjugate gradient method.
        
    :param x0 : vector
        The starting guess for the solution.
        
    :param max_iter : integer
        Maximum number of iterations. Iteration will stop after max_iter 
        steps even if the specified tolerance has not been achieved.
        
    :param tol : float
        Tolerance to achieve. The algorithm will terminate when either 
        the relative or the absolute residual is below tol.
        
    :var    r0 : vector
                 Initialization stores the gradient of x0 * (-1)
    
    :var    d  : vector
    
    :var    a  : float
                 Iteratively computes the vector of x

    :var    ri : vector
                 Iteratively stores the value norm of x ( or a), 
                 used to check for the convergence
    
    :var    x  : vector 
                 Stores the solution for the next iteration 
                 iteratively
    
    :sym    lmda  : symbol 
                 Used to compute the lamda value using 
                 Goldstein Armijo criteria
                 
    """
    x = np.matrix(x0)
    r0 = grad_vector(x0) * (-1)
    d = r0
    alpha = 0.5

#   Iterations:   
    for i in xrange(max_iter):
        
        #compute the lamda value by line search
        lmda = Symbol('lmda',real = True)
        # function value at x0
        f_x0 = func(x0[0],x0[1],x0[2],x0[3])
        # The Goldstien-Armijo criteria for the lambda selection
        rhs = f_x0 + np.dot(grad_vector(x0),d.T)*(alpha)*(lmda)
        lhs = func(x0[0] + lmda*d.item(0),x0[1] + 
                   lmda*d.item(1),x0[2] + lmda*d.item(2),x0[3] + 
                   lmda*d.item(3))
        # solver for the lamda value from quadratic inequality
        try:
            lmda_value = max(solve(lhs-rhs,lmda))
        except ValueError: 
            pass
        
        # line search method ends there
        x = x + np.multiply(d,lmda_value)
        a = np.array([x.item(0),x.item(1),x.item(2),x.item(3)])
        print "iteration: ",i, "r(i): ",
        round(math.sqrt(grad_vector1(a)[0]**2 + grad_vector1(a)[1]**2 + 
              grad_vector1(a)[2]**2  + grad_vector1(a)[3]**2),5)
        if math.sqrt(x.item(0)**2 + x.item(0)**2 + 
                      x.item(0)**2  + x.item(0)**2) < tol:
            print "\nConverged Successfully in iterations :",i
            print "The result of vector x:"
            return x
            break
        b = float((grad_vector(a)*grad_vector(a).T)/(
                  grad_vector(x0)*grad_vector(x0).T))
        d = grad_vector(a) * (-1) + d * b
        x0 = a
    return x

**10 Iterations**

In [14]:
non_linear_conjugate_gradient(x0, tol = 1.0e-8, max_iter = 10)

iteration:  0 r(i):  2.27121
iteration:  1 r(i):  0.8224
iteration:  2 r(i):  0.44142
iteration:  3 r(i):  0.18847
iteration:  4 r(i):  0.09744
iteration:  5 r(i):  0.0449
iteration:  6 r(i):  0.02238
iteration:  7 r(i):  0.01097
iteration:  8 r(i):  0.00506
iteration:  9 r(i):  0.00261


matrix([[1.10646976973941, 0.743843588792184, 0.610066169803682,
         0.552504690335135]], dtype=object)

**2 Iterations**

In [15]:
non_linear_conjugate_gradient(x0, tol = 1.0e-8, max_iter = 2)

iteration:  0 r(i):  2.27121
iteration:  1 r(i):  0.8224


matrix([[1.08908941845879, 0.764107786237180, 0.667004995370713,
         0.631226288110453]], dtype=object)

**Comments on Results**:
- No. of Iterations :** 10 iterations**
- Using the initial guess as **[1,1,1,1]** approxiamated value of x is **[1.106, 0.744, 0.61,0.55]**

- After **2 iterations**, the value of **r(i) = 0.8224**,value of x is **[1.089, 0.764, 0.667,0.631]**


**Comments on Method**:
- The evaluation of the non - linear function is done using n = 4.
- The alpha value is chosen to be **0.5**
- The lambda value is calculated using the maximum of the two roots from the quadratic inequality:
  - of the Goldstein - Armijo criteria.
- Please note that the lambda value can take any value between the roots, due to the inequality( lambda term to the L.H.S)
